### PR TITLE
Push docker image to dockerhub 

### PR DESCRIPTION
**Pull request by Spellchaser**
Closed #75 with a push by accident sorry @nsidhaye.
I rebased ontop of master. The travis check should fail now that the Dockerfile was moved

### CODE REVIEW

**Comment by Spellchaser** [comment]
If you want to publish the ubuntu or the alpine distro, I'll switch the default.

File: ubuntu/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,12 @@
+FROM golang:latest as builder
+
+RUN go get github.com/asobti/kube-monkey
+WORKDIR $GOPATH/src/github.com/asobti/kube-monkey
+RUN make build
+
 FROM ubuntu
```
Comment (nsidhaye) [inline ubuntu/Dockerfile]: I was not sure about changing default Dockerfile to ubuntu/Dockerfile. 

If you are golang developer then you might hate to download another docker image for golang which is not necessary as golang developer having all necessary compile & run time environment available. 

It makes sense to have different golang image if you do not have golang environment which is case in dockerhub environment.  As we are building docker image in dockerhub's environment we have to use golang env. to build image.

I think old Dockerfile is more deverloper friendly who suppose to have golang environment and able to make

File: ubuntu/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,12 @@
+FROM golang:latest as builder
```
Comment (nsidhaye) [inline ubuntu/Dockerfile]: I was not sure about changing default Dockerfile to ubuntu/Dockerfile. 

If you are golang developer then you might hate to download another docker image for golang which is not necessary as golang developer having all necessary compile & run time environment available. 

It makes sense to have different golang image if you do not have golang environment which is case in dockerhub environment.  As we are building docker image in dockerhub's environment we have to use golang env. to build image.

I think old Dockerfile is more developer friendly who suppose to have golang environment and able to execsute make locally.

Comment (nsidhaye) [reply ubuntu/Dockerfile]: Yes that's definitely benefit.

**Comment by asobti** [comment]
I can't actually test this as I don't have the version of docker required for multi-stage builds. I think it might be a good idea to continue supporting single-stage builds in addition to multi-stage for now.

Basically have make targets: `alpine`, `ubuntu` and `ubuntu-compat`. The last of which uses a single stage build

**Comment by asobti** [comment]
@Spellchaser Take a look at https://github.com/asobti/kube-monkey/commit/c5a66540e4380f7d20019578d5af62fcf66434ae

If this looks good, I can merge it into this PR

**Comment by Spellchaser** [comment]
So that would use the local golang build and ubuntu/alpine will be purely for release? 

I think there's several things to consider here.
 * How easy is it for a new contributor to pickup the project?
 * Do they have to have golang installed?
 * Are they willing to download an image?
 * Should the default build with the latest golang?
 * Should the Dockerfiles COPY instead of go get to be able to build the local changes?
 * Can we avoid duplication in Dockerfiles?

**Comment by asobti** [comment]
`ubuntu-compat` will use the local golang to build, correct. It only needs to be used by those running pre Docker17 versions.

With regards to contributors, I fail to see how one could contribute without having golang locally. At the very minimum, I would expect contributors to run tests and `gofmt` locally before submitting a PR.

Default, in my opinion, should only build with a version of Golang that this project supports. For example, I just merged https://github.com/asobti/kube-monkey/commit/be6f8b24608a0e125200737a66ed3e09c37a55ef, so as of now, we only support Go 1.10 and 1.11. 

Copy seems more appropriate. It makes testing local changes a lot easier.

**Comment by Spellchaser** [comment]
Yeah the compact works. LGTM, if I have any thoughts later on I'll make a PR

**Comment by asobti** [comment]
I do think we should modify all Dockerfiles to *not* use the latest golang version.

As for keeping Dockerfiles DRY, I'd love that but I'm not sure how we might do that.

**Comment by Spellchaser** [comment]
Voila! Docker args :) 

Why is it ubuntu-compat vs ubuntu-compact or local-build?

I think there's still a bit of work. It's confusing for someone new to run make containers but not have it include local changes.

**Comment by asobti** [comment]
`ubuntu-compat` for `compatibility` mode. 

And ugh, yeah - I didn't realize the ubuntu and alpine Docker builds are doing a `go get`. They should definitely mount the local code and build based on that so local changes are included.

**Comment by asobti** [comment]
@Spellchaser Also getting this error when running `make ubuntu-compat` with Docker 1.12

```
Removing intermediate container 6cdeabc1aafe
One or more build-args [GOLANG_VER] were not consumed, failing build.
make: *** [ubuntu-compat] Error 1
```

**Comment by asobti** [comment]
@Spellchaser Taking a step back, what does this PR buy us? I feel like we're building this fairly complex system to make containers without any tangible benefits. I don't think we need multi-stage builds.

The way I see it, there are two roles users fall into:

Developers - Developers will want to build containers with their local changes, but it is also reasonable to assume (and require) developers will have the goland toolchain available to them to develop, run tests etc. In this case, they can run `make container` which will build locally, and then build the container.

Users - If someone is just interested in using kube-monkey, and does not want to build it and doesn't have golang, they should just use the image from DockerHub.

**Comment by Spellchaser** [comment]
Nope, we don't. It's just the new fancy way to do docker :)

Makes sense to me. I would keep the alpine Dockerfile and keep the ubuntu one default. This way if someone wants to create the minimal build, we don't have to throw away the work and make them translate the `yum` to `apt-get`.

**Comment by Spellchaser** [comment]
Maybe we should pin the ubuntu version as well? It's basically pulling `latest`

```
testbox:~/go/src/github.com/asobti/kube-monkey(dockermaster) $ cp alpine/Dockerfile .

testbox:~/go/src/github.com/asobti/kube-monkey(dockermaster) $ make multi-stage
Building multi-stage container with ' --build-arg http_proxy=<SNIP> --build-arg <SNIP>'
Sending build context to Docker daemon  149.5MB
Step 1/9 : ARG GOLANG_VERSION
Step 2/9 : FROM golang:${GOLANG_VERSION:-1.10} as builder
1.10: Pulling from library/golang
05d1a5232b46: Already exists 
5cee356eda6b: Already exists 
89d3385f0fd3: Already exists 
80ae6b477848: Already exists 
94ebfeaaddf3: Already exists 
e132030a369d: Pull complete 
c67c5750c788: Pull complete 
Digest: sha256:a325153f1f1e2edb76a76ad789aff172b89dd6178e8f74d39ef87c04d87d1961
Status: Downloaded newer image for golang:1.10
 ---> a4afc24299ee
Step 3/9 : COPY . $GOPATH/src/github.com/asobti/kube-monkey
 ---> 8e51bc539659
Step 4/9 : WORKDIR $GOPATH/src/github.com/asobti/kube-monkey
Removing intermediate container 8c42230f939c
 ---> 2a095f4b6db6
Step 5/9 : RUN make build
 ---> Running in ac3825868194
Cleaning compiled kube-monkey binary:
rm -f kube-monkey
Checking gofmt:
find . -path ./vendor -prune -o -name '*.go' -print | xargs -L 1 -I % gofmt -s -w %
Warning golangci-lint not installed. Skipping linting
Installation instructions: https://github.com/golangci/golangci-lint
GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -o kube-monkey
Removing intermediate container ac3825868194
 ---> 648122fa1b33
Step 6/9 : FROM alpine
 ---> 11cd0b38bc3c
Step 7/9 : LABEL maintainer="asobti <ayushsobti@gmail.com>"
 ---> Using cache
 ---> 94627fdd335d
Step 8/9 : RUN apk update && apk upgrade && apk add --no-cache bash tzdata
 ---> Using cache
 ---> 5e1b31a11202
Step 9/9 : COPY --from=builder /go/src/github.com/asobti/kube-monkey/kube-monkey /kube-monkey
 ---> 2d36df4f933d
Successfully built 2d36df4f933d
Successfully tagged kube-monkey:v0.2.4
```

**Comment by amir-hadi** [comment]
There is one more very important thing about this pushing the image to Docker using CI or even using an public build (autobuild) on DockerHub: We can see what you are building.

Right now we are blind and need to trust that the Docker image you publish is exactly the the kube monkey we would expect.

**Comment by worldtiki** [comment]
Hi @nsidhaye! Sorry for the delay.
I'll try to review this as soon as I can. I have a separate pr to setup github actions. Once that's merged I'll add a new action to publish the docker image automatically.

**Comment by worldtiki** [comment]
Deprecated. Closing this pr.

**Pull request closed without merging**